query/src/main/scala/filodb/query/util/HierarchicalQueryExperience.scala

```diff
@@ -0,0 +1,234 @@
+
+import filodb.core.GlobalConfig
+import filodb.core.query.ColumnFilter
+import filodb.core.query.Filter.Equals
+import filodb.query.LogicalPlan
+
+final case class HierarchicalQueryExperience(isInclude: Boolean,
+                                             metricRegex: String,
+                                             metricSuffix: String,
+                                             tags: Set[String])
+object HierarchicalQueryExperience extends StrictLogging {
+
+  // Get the shard key columns from the dataset options along with all the metric labels used
+  lazy val shardKeyColumns: Option[Set[String]] = GlobalConfig.datasetOptions match {
+    case Some(datasetOptions) =>
+      Some((datasetOptions.shardKeyColumns ++ Seq( datasetOptions.metricColumn, GlobalConfig.PromMetricLabel)).toSet)
+    case None => None
+  }
+
+  // Get the allowed aggregation operators from the hierarchical config
+  lazy val allowedAggregationOperators: Option[Set[String]] = GlobalConfig.hierarchicalConfig match {
+    case Some(hierarchicalConfig) =>
+      Some(hierarchicalConfig.getStringList("allowed-aggregation-operators").asScala.toSet)
+    case None => None
+  }
+
+  // Get the allowed range functions from the hierarchical config
+  lazy val allowedRangeFunctions: Option[Set[String]] = GlobalConfig.hierarchicalConfig match {
+    case Some(hierarchicalConfig) =>
+      Some(hierarchicalConfig.getStringList("allowed-range-functions").asScala.toSet)
+    case None => None
+  }
+
+  // Get the allowed periodic series plans which have access to RawSeries from the hierarchical config
+  lazy val allowedPeriodicSeriesPlansWithRawSeries: Option[Set[String]] = GlobalConfig.hierarchicalConfig match {
+    case Some(hierarchicalConfig) =>
+      Some(hierarchicalConfig.getStringList("allowed-periodic-series-plans-with-raw-series").asScala.toSet)
+    case None => None
+  }
+
+  // Get the allowed parent logical plans for optimization from the hierarchical config
+  lazy val allowedLogicalPlansForOptimization: Option[Set[String]] = GlobalConfig.hierarchicalConfig match {
+    case Some(hierarchicalConfig) =>
+      Some(hierarchicalConfig.getStringList("allowed-parent-logical-plans").asScala.toSet)
+    case None => None
+  }
+
+  /**
+   * Helper function to get the ColumnFilter tag/label for the metric. This is needed to correctly update the filter.
+   * @param filterTags - Seq[String] - List of ColumnFilter tags/labels
+   * @param datasetMetricColumn - String - Metric ColumnFilter tag/label from the configured dataset options
+   * @return - String - ColumnFilter tag/label for the metric
+   */
+  def getMetricColumnFilterTag(filterTags: Seq[String], datasetMetricColumn: String): String = {
+    // get metric name filter i.e either datasetOptions.get.metricColumn or PromMetricLabel - We need to support both
+    // the cases
+    filterTags.find( tag => tag == datasetMetricColumn || tag == GlobalConfig.PromMetricLabel)
```

Comment:
Just being super cautious here to avoid any surprises that might result in an issue with just a tiny more cpu cost. 
But agree with you that logical plans just have `__name__` tag. 

let me know, and I can just update it to just `__name__` tag to keep our logic simple and also simplify code

Reply:
Sync'd offline on this 👍 we are keeping it as it is